[Woody] 미션 3 : 체스판 초기화

jihye-woo opened the pull request:
안녕하세요 Woody 입니다! 미션 3 PR 드립니다!

---

### 🤦‍♀️ 고민했던 부분
#### 1.  BoardTest - print()
 - 요구사항의 '흰색 Pawn과 검은색 Pawn을 생성해 저장하는 List를 추가해 구현'을 맞추면서도 전체 보드를 어떻게 표현할 수 있을까에 대한 고민이 있었습니다. List에서는 보드 위치에 대한 정보가 없기 때문에 두 개의 PawnList만 가지고 전체 보드를 출력하기 힘들다고 생각했습니다. 
 - ▶ 그래서 **Pawn[][] board** 를 Board 클래스 필드에 선언 · 구현했습니다.
 -  (+) 폰의 존재 유무와 위치까지 한 번에 파악할 수 있어서 적절하다고 생각했습니다! 사이즈가 정해져 있어서 출력할 때도 편리했구요!
 - (-) 결국 pawn을 저장하는 곳이 board와 pawn List 두 군데라서 중복으로 저장을 하기 때문에 좋은 방법인지는 잘 모르겠습니다.
 
 #### 2. Color enum
 - 아직까지는 enum이 필요 없다는 코멘트가 있었지만 그래도 color에 종속적인 representation이라는 상태가 새로 추가되면서 enum을 고려하지 않을 수 없었습니다. color에 관련된 enum을 추가하면서 함수의 인자 수가 줄었고, 늘어나던 static 상수의 수를 줄일 수 있었습니다. 그리고 Color enum을 인자로 가지는 메소드는 color 값에 대한 validation을 따로 할 필요가 없어서 예외 처리를 해주던 코드를 제거할 수 있었습니다.
 - 다만 아래와 같이 가독성이 떨어지는 코드가 생겨서 이후 로직 혹은 네이밍을 수정해야할 것 같습니다.
 -  ```java
      // 너무 컬러컬러컬러 하다, 네이밍 때문인가
      Color color = Color.color(colorName);
      ```

---

### 🤹‍♀️ 구현 및 수정 내용
⁕ 요구사항에 있는 내용은 제외하고 특별히 제가 더 구현하거나 수정한 내용만 적어놓았습니다.
⁕ 다 읽어보지 않으셔도 괜찮습니다! 코드 이해에 도움이 될까싶어서 적어보았습니다!

#### 1. Game 클래스
  - do-while 을 이용해서 사용자 입력을 연속적으로 받는다
        
#### 2. CommandUtil 클래스
  - 사용자 입력과 관련된 static 메소드들 위치
  - printMenu() : 명령어 입력 설명 출력
  - readCommand() : 사용자 명령을 받아서 게임을 종료할지 계속할지 결정
  
#### 3. Board 클래스
  - 필드 변수로 `List<Pawn> whitePawns`, `List<Pawn> blackPawns`, 그리고 `Pawn[][] board` 선언
  - initialize() : Pawn 하나 생성 시, pawn 리스트에 한 번, board에 한 번 총 두 번을 저장
  - getPawnsResult() : getWhitePawnsResult()와 getBlackPawnsResult() 중복 로직 제거
  - getPawnListByColor() : 인자로 color 정보를 받아 적절한 pawn 리스트(whitePawns 혹은 blackPawns)를 리턴한다.

#### 4. Pawn 클래스
 - color 상태 정보 타입 : 문자열 → Color enum
 - 생성자 인자는 color 문자열 하나만 - Pawn(String color)
    - representation은 Color enum을 통해 가지고 옴
 
#### 5. Color enum 클래스
 - 색에 대한 정보와 색에 따른 representation 상태를 함께 관리
 - color 문자열 정보와 represent 함수를 필드 변수로 가짐 ([Java Enum 활용기 참고](https://woowabros.github.io/tools/2017/07/10/java-enum-uses.html))

jihye-woo committed:
docs : 미션3 체크리스트 추가

  > 미션3.md 추가

  > checklist 디렉토리 생성
    - 미션 체크리스트 관련 마크다운 파일 위치

jihye-woo committed:
feat : PawnTest 테스트 메소드 기능 추가

  > Pawn의 representation 테스트 추가
    - color와 함께 representation 테스트

  > 테스트 실행 후 test failed 확인

jihye-woo committed:
feat : Pawn 클래스 representation 추가

  > Pawn 클래스 변수
    - representation을 나타내는 char 타입 상수
    - WHITE_REPRESENTATION = 'p'
    - BLACK_REPRESENTATION = 'P'

  > Pawn 클래스 - representation 상태 추가
    - (char 타입)
    - 필드 추가
    - 생성자 매개변수 추가
    - getRepresentation() 구현

  > 테스트 실행 후 test passed 확인

jihye-woo committed:
feat : BoardTest 보드 초기화 테스트 메소드 추가

  > 테스트 메소드 initialize() 구현
   - 흰색 Pawn 열(row 6)의 결과가 "pppppppp" 인지 검증
   - 검은색 Pawn 열(row 1)의 결과가 "PPPPPPPP"인지 검증

  > 테스트 실행 후 test failed 확인

jihye-woo committed:
edit : Board 클래스 pawns List, 색깔별로 분리

  > List 타입 whitePawn 과 blackPawn
    - pawns 대신 같은 타입(List<Pawn>)으로 whitePawn와 blackPawn를 필드에 추가

  > getPawnListByColor() 구현
    - 인자로 받은 color에 맞는 List 리턴

  > pawns 삭제로 영향 받는 메소드 수정
    - findPawn(), color 인자 추가
    - size(), whitePawn와 blackPawn 사이즈 리턴
    - add(), getPawnListByColor()를 통해 적절한 List를 선택 후 추가

jihye-woo committed:
feat : Board 클래스 initialize() 구현

  > initialize()
    - Pawn 색깔별로 8개씩 생성하고 각각 whitePawns, blackPawns에 추가

jihye-woo committed:
feat : Board 클래스 PawnsResult getter 구현

  > getWhitePawnsResult()
    - whitePawns가 있으면 representation 없으면 '.'으로 표현
    - StringBuilder로 구현해서 string 리턴

  > getBlackPawnsResult()
    - getWhitePawnsResult()와 동일한 로직

jihye-woo committed:
feat : BoardTest 클래스 print() 구현

  > BoardTest::print
    - Board::print 를 테스트 하는 메소드
    - 보드를 초기화 한 뒤 출력된 결과에 폰들이 적절한 위치에 자리하고 있어야 한다.

  > 테스트 실행 후 test failed 확인

jihye-woo committed:
feat : Board 클래스 print() 구현

  > print()
    - 보드판 전체를 출력
    - board를 순회하며 pawn이 있으면 representation 출력, 없으면 '.' 출력

  > board 선언 타입 변경
    - List<Pawn> → Pawn[][]
    - List<Pawn>로 선언하면 pawn이 없는 상태까지 표현하기 힘들다
    - Pawn[][]로 선언하면 특정 위치에 pawn 존재 유무 파악 쉽다

jihye-woo committed:
edit : Board 클래스 initialize() 수정

  > Board::initialize
    - board에 pawn을 넣어주는 작업을 추가

  > BoardTest::print
    - assertThat 구문에서 expectedResult와 actualResult 위치 스왑

  > 테스트 실행 후 test passed 확인

jihye-woo committed:
feat : 메인 클래스 Game 구현

  > 시작과 종료 기능
    - Scanner로 사용자 입력 값 받는다.
    - 사용자 입력 명령을 수행한다. (시작 혹은 종료)
    - 종료 명령을 받을 때, 프로그램이 종료된다.

jihye-woo committed:
feat : CommandUtil 클래스 구현

  > 사용자 명령 처리 책임
   - printMenu() : 입력 가능한 명령어와 설명 출력
   - readCommand() : 입력받은 명령어를 해석

jihye-woo committed:
refactor : Board 클래스 getPawnsResult()로 중복 제거

  > getWhitePawnsResult()와 getBlackPawnsResult() 중복
    - getPawnsResult()으로 공통 로직 이동

jihye-woo committed:
feat : Color enum 클래스 구현

  > Color enum
    - color와 representation 상태를 한 번에 관리

  > color()
    - 인자로 받은 color 문자열과 매핑되는 Color enum을 리턴

  > representation()
    - 색에 따라 다른 representation 정보를 들고옴

jihye-woo committed:
edit : Pawn 클래스, Color enum 적용

  > Pawn 클래스
    - pawn을 대표하는 PAWN_REPRESENTATION 상수만 구현하고 나머지 color, representation 관련 static 상수들 제거
    - Pawn 생성자 인자, color 문자열 하나만

  > Color enum 클래스
    - color 문자열 이름을 필드에 추가

jihye-woo committed:
edit : PawnTest 클래스, Color enum 적용

  > PawnTest 클래스 전체 메소드 수정
    - Pawn 클래스 수정으로 인해 컴파일 에러난 코드
      → 수정 완료

  > 테스트 실행 후 tests passed 확인

jihye-woo committed:
edit : Board 클래스, Color enum 적용

  > Board 클래스, color 관련 메소드들
    - color 문자열 대신 Color enum 사용하게 수정

jihye-woo committed:
edit : BoardTest 클래스, Color enum 적용

  > BoardTest 클래스, color 정보 사용하는 메소드 수정
    - Board 클래스 수정으로 인해 컴파일 에러난 코드
    → 수정 완료

  > 테스트 실행 후 tests passed 확인

jihye-woo committed:
edit : Main 삭제, honux 패키지명 변경

  > 기존의 Main 클래스 삭제
    - 사용하지 않는 클래스 삭제

  > 패키지명 honux → woody
    - 죄송해용 바꾸겠습니다.

jihye-woo committed:
edit : 체스게임 시작 종료 기능 수정

  > 보드 시작 기능
    - 체스판의 현재 상태를 콘솔 화면에 출력

  > 보드 종료 기능
    - 보드 종료와 함께 종료 메세지 출력

jihye-woo committed:
refactor : Pawn 열에 대한 상태는 Pawn 클래스로 이동

  > WHITE_PAWN_ROW과 BLACK_PAWN_ROW
    - Board 클래스 → Pawn 클래스로 이동

imjinbro commented on src/main/java/net/woody/chess/Board.java:
```diff
@@ -0,0 +1,83 @@
+
+    private static final int BOARD_LENGTH = 8;
+
+    private final Pawn[][] board = new Pawn[BOARD_LENGTH][BOARD_LENGTH];
```
Pawn의 위치를 표시해주기 위해서 배열을 가지도록 한 것 같은데 조금 더 쉬운 방법이 없을까요?
Pawn 위치가 변경될 때마다 board 2차원 배열 요소에 할당된 값을 변경해야할 것 같아요

imjinbro commented on src/main/java/net/woody/pieces/Color.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    public Character representation(char pieceFirstLetter) {
+        return represent.apply(pieceFirstLetter);
```
자바8에 대한 공부의 목적으로 람다식을 저장해두었다가 호출하는건가요?

imjinbro replied:
아니시라면 매번 불필요한 호출보다는 정확한 값을 저장해두는 것이 이후 말에 따른 문자가 변경되었을 때 빠르게 적용할 수 있을 것 같네요
그리고 성능 상 크게 누수되는 것은 아니나 그래도 불필요한 메소드 호출 하는 것은 지양하는 것이 좋습니다 :)

jihye-woo replied:
### ✨ 람다식을 저장해두었다가 호출하는 이유
공부 목적이라기 보다는! Color와 representation의 대소문자 상태가 연관이 있다고 생각을 했습니다. 그래서
- color 값
- representation 대소문자 상태

이 두 가지를 함께 묶고 싶어서 Color enum 안에 위와 같은 람다식을 저장하게 되었습니다.

### 😮 덕분에 불필요한 연산이 반복되는 부분을 발견했습니다!
말씀을 듣고보니 Board 클래스 안에 getPawnsResult() 메소드에 representation()로 인한 연산이 불필요하게 많이 이루어지고 있네요! 얼른 수정해야겠습니다 감사합니다🙇‍♀
```java
private String getPawnsResult(Color color) {
        StringBuilder sb = new StringBuilder();
        for (Pawn pawn : getPawnListByColor(color)) {
            char represent = (pawn == null) ? '.' : color.representation(Pawn.PAWN_REPRESENTATION); // 👈👈
            sb.append(represent);
        }
        return sb.toString();
    }
```
위에서 불필요하게 Color::representation 이 호출되는 부분은 아래와 같이 수정하려고 합니다!
```java
char represent = (pawn == null) ? '.' : pawn.getRepresentation(); // 👈👈
```

### 🙋‍♀️ BUT 보통은 representation() 메소드가 자주 호출되지 않습니다!
'정확한 값을 저장' 해두는 방식을 말씀해주셔서 제 코드에 대해서 조금 더 자세히 설명드리고 싶습니다!

위의 상황을 제외하고 Color enum의 representation() 메소드를 호출하는 경우는 Pawn이 생성될 때 Pawn 생성자 안에서 아래와 같이 호출할 때 뿐입니다!
제가 구현해놓은 Pawn 클래스 필드에 `representation 변수`를 가지고 있어서 생성된 Pawn이 representation 상태를 가져오고 싶을 때는, Color의 representation()을 호출하지 않고, **Pawn 객체 안에 이미 생성된 representation을 가지고 옵니다**!
```java
public class Pawn {
    public static final char PAWN_REPRESENTATION = 'p';
    private Color color;
    private char representation; // 👈👈

  // ....

    public Pawn(String color) {
        this.color = Color.color(color);
        this.representation = this.color.representation(PAWN_REPRESENTATION); // 👈👈
    }
```

⁕⁕ 혹시 제가 코멘트를 잘못 이해하고 있다면 말씀해주세요! 감사합니다🙇‍♀️

imjinbro commented on src/main/java/net/woody/pieces/Pawn.java:
```diff
@@ -0,0 +1,31 @@
+package net.woody.pieces;
+
+public class Pawn {
```
1. Pawn이 Color를 가지는 것으로 하는건 어떨까요?
p와 P처럼 대소문자 구분이 아닌 전혀 다른 문자로 구분할 경우에는 조금 어려워질 듯 하네요

2. Board 클래스에 남겨둔 댓글에 대한 약간의 힌트를 드리자면 Pawn가 위치값을 가지면 어떨까요?

jihye-woo replied:
### 1. Pawn이 Color를 가지는 것으로 하는 건 어떨까요?
```java
public class Pawn {
    public static final char PAWN_REPRESENTATION = 'p';
    public static final int WHITE_PAWN_ROW = 1;
    public static final int BLACK_PAWN_ROW = 6;

    private Color color; // 👈👈
    private char representation;

    public Pawn() {
        this(Color.WHITE.colorName());
    }
...
}
```
- 현재 Pawn 클래스는 위와 같이 Color 클래스를 가지고 있습니다! 혹시 제가 잘못 이해하고 있다면 **'Pawn이 Color를 가지는 것'** 이 무엇을 말씀하시는건지 다시 한 번 여쭤봐도 될까요?

### 2. Pawn이 위치값을 가지면 어떨까요?
그렇군요! Array를 최대한 사용하지 않고 ArrayList를 사용하라는 요구사항을 반영하기 위해서는 위치값 정보를 따로 저장해두는게 좋을 것 같습니다!

** 글이 너무 길어져서 죄송합니다🙇‍♀️ 짧고 이해하기 쉽게 말씀드리고 싶었는데 쉽지가 않네요😂

imjinbro replied:
오호 제가 그 전에 코드에서 Color를 가지고 있지 않은 것으로 봐서 코멘트를 드렸는데 있었네요
코드가 그 사이에 뭔가 변경이 많아 보이네요
Board에 있던 2차원 배열도 없어졌고...

Pawn이 Board에서 움직인다
-> Pawn(체스말)이 x,y 위치값을 가지고 있으면 Board가 2차원 배열을 가지면서 말이 어딨는지 표시를 안해도 될것 같다

는 생각에 드렸던 피드백입니다!

객체지향적으로 사고하기를 연습하시면 더 나은 코드가 될 것입니다~!


그리고 미션 단계별로 브랜치를 따고 진행하시는거죠?

jihye-woo replied:
#### 브랜치
넵넵 단계별로 브랜치 따서 진행하고 있습니다! 다만, 커밋이 좀 길어져서..😂😂 믿기지 않으시겠지만 현재 PR에는 미션 3와 관련된 커밋만 있습니다(jihye_initBoard라는 이름으로요)!

#### 위치 값
주신 피드백 덕분에 위치 값에 대해서도 생각해보게 되었습니다! 의도하신 부분과는 살짝 다르게 구현하긴 했지만요..!😅

#### Board의 2차원 배열
(모두 다 이해가 되는 내용은 아니지만) 이펙티브 자바 아이템 28을 참고하니까 배열보다는 리스트가 더 안전하다는 생각을 하게 되었습니다! 그런 이유로 이번 미션에서 배열 대신 리스트를 사용하라고 하신게 아닐까 싶어서 변경하게 되었습니다..ㅎㅎ

imjinbro reviewed (changes requested):
출력 관련된 요구사항이 많다보니 Board 구현에 많은 노력을 기울이신 것으로 보입니다
출력을 조금 덜 신경쓰는 쪽으로 코드를 변경하실 수 있을 것 같아 코멘트 남겼습니다

jihye-woo committed:
refactor : Color::representation로 발생하는 비효율적인 연산 제거

  > Board 클래스 getPawnsResult()
    - color.representation(Pawn.PAWN_REPRESENTATION); 대신
    - pawn.getRepresentation() 사용
    - Pawn 객체에 이미 만들어진 representation 데이터 사용

jihye-woo committed:
feat : Rank 클래스 구현

  > 보드의 체스말 위치 정보를 나타내기 위해 구현
    - rank number 정보와 해당 rank의 폰들을 리스트로 가진다.

jihye-woo committed:
feat : Board 클래스 List<Rank> board 필드값 선언

  > whitePawns와 blackPawns - 대신 → board
  - 폰을 담는 곳을 board로 통일
  - 색 대신 rank로 구분

  > Pawn 클래스의 클래스 변수명 변경
    - rank 개념 적용 이후, row보다는 rank라는 명칭이 더 적절
    - WHITE_PAWN_ROW → WHITE_PAWN_RANK
    - BLACK_PAWN_ROW → BLACK_PAWN_RANK

jihye-woo committed:
feat : Rank 클래스 기능 추가

  > 메소드 추가 및 수정
    1. getPawns() 제거
      - 외부에서 pawns를 직접 수정 X
    2. find()
      - file(column) 정보를 가지고 pawn을 찾을 수 있음
    3. add()
      - Board 클래스의 add() 메소드 이름과 통일하기 위해 set -> add 이름 변경
      - 현재 기능에 맞추기 위해서 Pawn을 인자로 받아 추가하도록 수정
    4. toString()
      - 현재 기능에 맞추기 위해서 간단하게 수정

jihye-woo committed:
edit : Board 클래스 수정 - Rank 기능 추가에 의해

  > Rank 타입 추가에 따른 메소드 수정
    - 생성자 호출 시 board arraylist 빈 Rank로 초기화
    - 전체 메소드에 영향 → rank 관련 부분 수정

jihye-woo committed:
edit : Pawn 클래스, 클래스 변수 값 변경

  > 색깔 별 rank 값 수정
    - BLACK_PAWN_RANK, WHITE_PAWN_RANK 값 swap

jihye-woo committed:
edit : BoardTest 수정 - Rank 기능 추가에 의해

  > 헬퍼 메소드 getPawnRank() 추가
    - Pawn의 rank를 찾는 기능

  > Board 클래스의 findPawn() 시그니처 변경에 따른 수정
    - rank와 file 값을 통해 폰을 찾도록 수정

  > 테스트 실행 후 tests passed 확인

jihye-woo committed:
edit : Rank 클래스의 rank 변수 제거

  > Rank 클래스의 rank 변수가 현재 사용되고 있지 않음
   - 불필요한 rank 변수 제거

jihye-woo commented:
### 위치 정보를 위해서 Rank 클래스 추가
Pawn이 위치값을 가지면 좋겠다는 코멘트를 주셔서 보드의 행(row) 역할을 하는 `Rank 클래스`를 만들었습니다.
Rank가 각 rank마다 존재하는 Pawn list들을 관리합니다.
(이후 몇 번 rank 어느 file의 체스말을 이동하는 명령을 수행하기 위해서는 이런 구조가 적절하다고 생각했습니다) 

이 과정에서 `Board 클래스`에서 Color 기준으로 동작했던 메소드들이 **Rank 기준으로 동작**하게 끔 수정하였습니다.

가장 크게 바뀐 부분은 `List<Pawn> whitePawn`과 `List<Pawn> blackPawn`의 역할을 `List<Rank> board` 가 대신 수행한다는 점입니다. 

이로 인해 전체 메소드가 외형적으로는 조금씩 바뀌게 되었는데,  Rank 기준으로 바뀌었다는 것 빼고 기능은 이전과 동일합니다. 

### 간결한 출력 형태
그리고 Rank 클래스 덕분에 Board 클래스의 출력 메소드도 이전보다는 더 간결한 형태로 구현할 수 있게 되었습니다.

⁕⁕ 기존에 작성해놓았던 테스트 메소드는 모두 통과하였습니다.

jihye-woo committed:
edit : 커맨드 메뉴 출력 문구 수정 - minor issue

  > CommandUtil 클래스의 printMenu() 수정
  "not case-sensitive" → "case-insensitive"

jihye-woo committed:
refactor : 패키지 분리 - minor issue

  > CommandUtil 클래스
    - utils 패키지로 이동

ghojeong commented on src/main/java/net/woody/chess/Board.java:
```diff
@@ -0,0 +1,72 @@
+
+    public Rank getRank(int rank) {
+        if (rank < 0 || board.size() <= rank) {
+            throw new ArrayIndexOutOfBoundsException("Rank number " + rank + " is out of range!");
```
범위를 벗어났을 때의 예외처리 매우 👍  입니다.
저도 이것때문에 고민 많이 했었는데 보기 좋네요

jihye-woo replied:
흐힣 감사합니다!!😊😊

ghojeong commented on src/test/java/net/woody/pieces/PawnTest.java:
```diff
@@ -30,9 +32,10 @@ void createRedPawn() {
     }
 
-    private void verifyPawn(String color) {
+    private void verifyPawn(final String color) {
```
파라미터에 굳이 final 을 붙이시는 이유가 있나요? 
이러면 무엇이 좋은지, 의도가 정말로 궁금해서 여쭈어봅니다.

jihye-woo replied:
사실 이 부분은 예시 코드에  나와있던 부분을 그대로 사용한거였는데, 덕분에 한 번 더 생각해보게 되었습니다😳👍
제가 생각하는 이유는 아래 두 가지입니다!

1. 파라미터에 final을 붙여주는 이유가 **인자로 받은 값을 변경하지 않는다**는 것을 좀 더 명확하게 명시할 수 있다.
   - 코드를 공유하는 다른 개발자들도 코드를 일일히 뜯어보지 않아도 인자 값이 변경되면 안된다는 것을 확실하게 알 수 있을 것 같아요!
  
 2. 재선언하면 안되는데 재선언을 했을 시, **컴파일 타임에 미리 확인**할 수 있다.
    - 사실 final이 없어도 개발자 재량껏 color를 변경하지 않는 로직을 작성하지 않는다면 문제가 없겠지만 혹시나 리펙토링 과정에서 깜빡하고 color를 재선언한다던지 그런 불상사를 미연에 방지할 수 있을 것 같아요!
    
 사실 말은 이렇게 했지만 저도 의식적으로 파라미터에 final 선언을 한 적이 없어서 개인적인 의견으로 참고만 해주세요! 예리한 질문✨ 감사드립니다🙇‍♀️

ghojeong commented on src/main/java/net/woody/pieces/Pawn.java:
```diff
@@ -0,0 +1,31 @@
+        this.representation = this.color.representation(PAWN_REPRESENTATION);
+    }
+
+    public String getColorName() {
+        return color.colorName();
+    }
+
+    public Color getColor() {
+        return color;
+    }
```
getColor 외에 getColorName 이 한개 더 존재 해야하는 이유가 있나요?
아마도 테스트 용도 때문에 getColorName 함수를 만드신 것 같은데,
`color.name().toLowerCase()` 를 통해 getColorName 을 지우는 것은 어떨지 이야기 해봅니다.
이것은 Color 를 String 값으로 받는 것을 싫어하는, 지극히 개인적인 취향으로 인한 피드백이라서
무시하셔도 될 것 같습니다.

jihye-woo replied:
안그래도 저도 저 부분에 대해서도 고민을 좀 했었습니다..! 연속적으로 참조하는 부분이 길어지면 가독성이 떨어질 것 같아서 피하고싶긴 했었거든요😫

그리고! 파이로 말씀하신거 듣고 아이디어가 떠올랐어요! 
Color enum 클래스에 toString을 오버라이드해주면 assert문에서는 아래와 같이 호출해줄 수 있겠더라구요!
```java
assertThat(pawn.getColor()).isEqualTo(color);
```
맞는 방법인지는 잘 모르겠지만 이런 식으로 수정해보려고 합니다! 
자세하게 짚어주셔서 감사합니다!!🙇‍♀️

ksundong replied:
이 부분은 어느정도 디미터 법칙과도 연관있어보입니다.
한 번 공부해보세요.

jihye-woo replied:
오옹  디미터 법칙 처음 들어봐요! 좋은 키워드 감사합니다😁 
연속적으로 호출해야할 일이 생겼을 때, 객체 사이의 결합도에 대해서도 고려해 봐야겠군요!

ghojeong replied:
저도 디미터 법칙 덕분에 알게되었습니다.
디미터라고 하면 어색한데,
데메테르 여신님 법칙 이라고 하니깐 확 와닿네요 ㅋㅋㅋ

ghojeong commented on src/main/java/net/woody/utils/CommandUtil.java:
```diff
@@ -0,0 +1,25 @@
+            case "start":
+                return true;
+            case "end":
+                System.out.println("bye~!💨💨");
```
ㅋㅋㅋ 종료 메시지 귀엽습니다.

jihye-woo replied:
감사합니다아💨💨

ghojeong commented on src/main/java/net/woody/utils/CommandUtil.java:
```diff
@@ -0,0 +1,25 @@
+        System.out.print("> ");
+    }
+
+    public static boolean readCommand(String command) {
```
좀더 좋은 함수명으로 지으면 어떨지 제안을 해봅니다.
"readCommand" 라는 이름만 들었을 때는, void 함수이거나, String 값을 리턴하는 함수처럼 생각이 되어서요.
일반적으로 boolean 값을 리턴하는 함수 앞에는 "is" 나 "has" 같은 질문형 접두사를 붙입니다.
반환 받은 true 와 false 가 무엇을 의미하는지 바로 알수 있게끔요.

가령 isGameOver() 같은 함수명이라면, true 일때 게임이 종료되어야 하고, false 일때 게임이 계속 되어야 한다는 것을 바로 알 수 있겠지요.

jihye-woo replied:
옷 그렇군요! 말씀하신 `isGameOver()` 마음에 들어요! boolean 값을 리턴하니까 이런 네이밍이 더 어울린다는 생각이 드네요 수정해야겠어요 감사합니다!ㅎㅎ

ghojeong commented on src/test/java/net/woody/pieces/PawnTest.java:
```diff
@@ -12,15 +12,17 @@
     void createWithDefaultPawn() {
         Pawn pawn = new Pawn();
-        assertThat(Pawn.WHITE_COLOR).isEqualTo(pawn.getColor());
+        assertThat(Color.WHITE).isEqualTo(pawn.getColor());
```
```suggestion
        assertThat(pawn.getColor())
            .isEqualTo(Color.WHITE);
```

호눅스님이 수업때 말씀하셨던 것 것처럼, actual 을 앞에두고, expected 를 뒤로 써주시는 게 좋습니다.
intelliJ 의 자동완성에서 보여주는 파라미터 명을 가능한 따라주시면 좋아요

jihye-woo replied:
완전 자세하게 봐주셔서 감사드려요! 
이 부분은 제가 제대로 확인하지 않았네요!! 수정하도록 하겠습니다 감사합니다😀😊

ghojeong commented on src/main/java/net/woody/pieces/Color.java:
```diff
@@ -0,0 +1,37 @@
+        this.represent = represent;
+    }
+
+    public static Color color(String color) {
+        switch (color) {
+            case "white":
+                return WHITE;
+            case "black":
+                return BLACK;
+            default:
+                throw new IllegalArgumentException("color should be white or black!");
+        }
+
+    }
```
```java
String str = "white";
Color.valueOf(str.toLowerCase());
```
를 하면 valueOf 함수만으로 충분히 color 함수를 대체할 수 있지 않을까 제안해봅니다.
코드 중에서 아래와 같은 상황이 발생하면 정말 혼돈스러울 것 같아서요.
```java
String color = "white";
...
Color color = Color.color(this.color);
```

jihye-woo replied:
`valueOf()` 수업시간에 보고서 사용할 생각은 못했네요! 
맞아요😳 너무 컬러컬러컬러해서 안그래도 고민이 많았는데 감사합니다!

jihye-woo replied:
그런데 예외 처리부분은 남기고 싶어서 다시 고민해보다가 `color()` 메소드 명을 `value()`로 바꾸기로 했습니다..ㅠ!!

ghojeong replied:
넵 훌륭하신 판단같습니다.

ghojeong commented on src/main/java/net/woody/chess/Game.java:
```diff
@@ -0,0 +1,23 @@
+        do {
+            CommandUtil.printMenu();
+            String command = sc.nextLine();
+            ongoing = CommandUtil.readCommand(command);
```
읽으면서 느끼지만, readCommand 를 했더니 ongoing 을 받아와진다는게, 네이밍이 매우 잘 안 읽힙니다.
ongoing 과 readCommand 가 서로 연관되게 다시 이름을 지어주시면 좋을 것 같습니다.

jihye-woo replied:
넵넵 다른 코멘트에 남겨주신대로 네이밍 수정하는게 좋을 것 같습니다! 감사합니다🙇‍♀️🙇‍♀️

ghojeong commented on src/main/java/net/woody/chess/Board.java:
```diff
@@ -0,0 +1,72 @@
+        Rank blackPawns = getRank(Pawn.BLACK_PAWN_RANK);
+        Rank whitePawns = getRank(Pawn.WHITE_PAWN_RANK);
+
+        IntStream.range(0, BOARD_LENGTH).forEach(i -> {
```
```java
for (int i = 0; i < BOARD_LENGTH; i++) {
}
```

대신 굳이 IntStream 을 사용하신 이유를 알 수 있을까요?

jihye-woo replied:
사실 작성할 때까지만 해도 어떤 방식이 여기서 더 적절할지 몰라서 개인적으로 더 편한 방식인 IntStream을 사용했습니다!
남겨주신 코멘트 보고 더 찾아보니까 람다식이 퍼포먼스 측면에서는 대개 for-loop보다 좋지 않기 때문에, 위와 같이 람다식으로 간결해지는 장점을 활용할 수 없을 때에는 for-loop을 사용하는 것이 더 좋을 것 같네요!

좋은 포인트 짚어주셔서 감사합니다!😊😊

참고 - [3 Reasons why You Shouldn’t Replace Your for-loops by Stream.forEach()](https://blog.jooq.org/2015/12/08/3-reasons-why-you-shouldnt-replace-your-for-loops-by-stream-foreach/)

ksundong replied:
`IntStream.range()`는 지양해주세요 ㅎㅎ

jihye-woo replied:
네엡! 지양해야겠습니다😂😂

ghojeong reviewed (commented):
와우 코드 정말 세심하게 짜셨네요 👍 
case-insensitive 하게 배려하신 점이나,
다양한 자바 문법을 활용해내신점이나 무척 인상 깊습니다.
피드백이라기 보다는, 읽으면서 궁금점이 생겼던것을 코멘트로 남겨보았습니다.

jihye-woo committed:
refactor : Color 클래스 수정

  > colorName 필드 삭제
    - toString() 오버라이드로 대체

  > 가독성 문제로 color() 메소드명 변경
    - 메소드명 value()로 변경

  > 영향 범위에 있는 생성자와 메소드들 수정

jihye-woo committed:
refactor : assert 메소드 변수 순서 변경

  > PawnTest 테스트 메소드 수정
    - createWithDefaultPawn()에서 expected 값과 actual 값 순서 변경

imjinbro commented:
수고하셨습니다!

ghojeong reviewed (approved):
👍 😆
